src/django_mysql/test/cases.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import unicode_literals
+
+import sys
+
+from django.apps import apps
+from django.db import connections, transaction
+from django.db.models.fields import AutoField
+from django.core.management.base import CommandError
+from django.core.management import call_command
+from django.test import TransactionTestCase
+from django.utils import six
+
+
+models_without_auto_inc = []
+
+
+def find_models_without_auto_inc():
+    all_models = apps.get_models()
+    for model in all_models:
+        fields = model._meta.fields
+        has_auto = False
+        for field in fields:
+            if isinstance(field, AutoField):
+                has_auto = True
+                break
+        if not has_auto:
+            models_without_auto_inc.append(model._meta.db_table)
+
+
+if not models_without_auto_inc:
+    find_models_without_auto_inc()
+
+
+class FasterTransactionTestCase(TransactionTestCase):
+    def _fixture_teardown(self):
+        # Allow TRUNCATE ... CASCADE and don't emit the post_migrate signal
+        # when flushing only a subset of the apps
+        for db_name in self._databases_names(include_mirrors=False):
+            # Flush the database
+            inhibit_post_migrate = (
+                self.available_apps is not None or (
+                    # Inhibit the post_migrate signal when using serialized
+                    # rollback to avoid trying to recreate the serialized data.
+                    self.serialized_rollback and
+                    hasattr(connections[db_name], '_test_serialized_contents')
+                )
+            )
+            connection = connections[db_name]
+            if "mysql" in connection.settings_dict['ENGINE']:
+                tables = connection.introspection.django_table_names(only_existing=True, include_views=False)
+                schema_name = connection.settings_dict['NAME']
+
+                # auto_increment is reliable, but table_rows isn't
```

Comment:
From what I can tell, this is true, but it'd be good if someone familiar with MySQL internals could confirm.

Reply:
It's true it's unreliable. Therefore I think this isn't safe in the general case, since a table with a non-auto-inc PK could be reported as empty when not, and then not truncated. You could do `select 1 from x limit 1` on those tables that returned 0. And even `union` together such queries, in batches, to fetch details for multiple tables in one batch, e.g. for two tables:

```
chainz@localhost [14]> (select 't' as has_rows from t limit 1) union (select 't2' as has_rows from t2 limit 1);
+----------+
| has_rows |
+----------+
| t2       |
+----------+
1 row in set (0.001 sec)
```

Reply:
Also `delete from` can be faster for low row counts, since it doesn't physically recreate the table on disk. With a single row table:

```
chainz@localhost [21]> truncate t2;
Query OK, 0 rows affected (0.005 sec)

chainz@localhost [22]> delete from t2;
Query OK, 0 rows affected (0.000 sec)
```

I think even this optimization would work, using estimated `row_count` - `delete from` for small tables, `truncate` for big ones.